openssl/src/ossl_param.rs

```diff
@@ -0,0 +1,216 @@
+//! OSSL_PARAM management for OpenSSL 3.*
+//!
+//! The OSSL_PARAM structure represents an array of generic
+//! attributes that can represent various
+//! properties in OpenSSL, including keys and operations.
+//!
+//! This is always represented as an array of OSSL_PARAM
+//! structures, terminated by an entry with a NULL key.
+//!
+//! For convinience, the OSSL_PARAM_BLD builder can be used to
+//! dynamically construct these structures.
+//!
+//! Note, that this module is available only in OpenSSL 3.* and
+//! only internally for this crate.
+
+use crate::bn::{BigNum, BigNumRef};
+use crate::error::ErrorStack;
+use crate::util;
+use crate::{cvt, cvt_p};
+use foreign_types::{ForeignType, ForeignTypeRef};
+use libc::{c_char, c_uint, c_void};
+use openssl_macros::corresponds;
+use std::ffi::CStr;
+use std::ptr;
+
+foreign_type_and_impl_send_sync! {
+    // This is the singular type, but it is always allocated
+    // and used as an array of such types.
+    type CType = ffi::OSSL_PARAM;
+    // OSSL_PARMA_free correctly frees the entire array.
+    fn drop = ffi::OSSL_PARAM_free;
+
+    /// `OsslParamArray` constructed using `OsslParamBuilder`.
+    /// Internally this is a pointer to an array of the OSSL_PARAM
+    /// structures.
+    pub struct OsslParamArray;
+    /// Reference to `OsslParamArray`.
+    pub struct OsslParamArrayRef;
+}
+
+impl OsslParamArrayRef {
+    /// Locates the individual `OSSL_PARAM` element identified by the key
+    /// in the `OsslParamArray` array and returns a reference
+    /// to it.
+    #[corresponds(OSSL_PARAM_locate)]
+    #[allow(dead_code)] // TODO: remove when when used by ML-DSA / ML-KEM
+    pub(crate) fn locate<'a>(&'a self, key: &CStr) -> Result<&'a OsslParam<'a>, ErrorStack> {
+        unsafe {
+            let param = cvt_p(ffi::OSSL_PARAM_locate(self.as_ptr(), key.as_ptr()))?;
+            Ok(OsslParam::from_ptr(param))
+        }
+    }
+}
+
+/// A reference to a single `OSSL_PARAM` structure inside
+/// an `OsslParamArray`.
+pub(crate) struct OsslParam<'a>(&'a ffi::OSSL_PARAM);
```

Comment:
This gives us a double indirection, which I don't think we want?

Reply:
OsslParam has a reference to an OSSL_PARAM, but then we return a reference to OsslParam